### ISSUE TITLE
With pirserver:: serialize_ reply method, but there is no corresponding deserialize. 

### ISSUE THREAD

**Issue by chasglim** [Microsoft/SealPIR#21]
With pirserver:: serialize_ reply method, but there is no corresponding deserialize. 
After transmission through the network, there is no way to deserialize it to reply on the client. 
In the main function, after the reply is serialized into serverstream, the serverstream is not used, but the reply is directly used in the following steps

**Comment by sga001** [comment]
Ah yes, it looks like we don't have deserialization for PIR reply, but it is probably similar to that of deserialize_query. I'll look into it over the weekend.

**Comment by chasglim** [comment]
Yes, I actually wrote some code before that by referring to deserialize_query, but there are always problems with client-side decoding, so I'll wait for your code.

**Comment by KinanBab** [comment]
Any updates on this? I also need the de-serialization for PirReply, but fortunately I was able to write my own (very simple) version that works fine for my use-case. If no one is looking into implementing a proper version of this, I could take a stab at it.

For the benefit of others, I will put below my simple de serialization which works for me.

I am writing something where the query and reply are both sent over a socket between client and server. The server uses `generate_reply(...)` to generate a `PirReply` and then calls `serialize_reply(reply, <std::stringstream>)` to serialize it into a string which I then send over the network socket.

On the client side, I read the exact serialized string (I know its size), say I store that string in a variable `std::string serialized_reply`. I then de-serialize it using:
```
std::stringstream istream(serialized_reply);
PirReply reply;
while (istream.rdbuf()->in_avail() > 0) {
  seal::Ciphertext c;
  c.load(context, istream);
  reply.push_back(std::move(c));
}

// context: the seal context initialized with parameters matching  SealPIRseal
// e.g. seal::SEALContext(seal::EncryptionParameters{seal::scheme_type::bfv}, true)
```

This works fine for me because (1) I am reading the serialized reply into a string, (2) I know its size, and (3) I am using `std::stringstream` which allows me to check if the entire string has been consumed yet. This code will not work when reading from a generic stream that may have many messages/replies streaming in continuously or if the exact string size is not known.

A proper implementation should probably calculate the size of the reply a priori (similar to the server's deserialize_query), and then load only the exact number of ciphers without needing to checking for the stream capacity.

**Comment by cnquang** [comment]
Me too. I also need the deserialization for the PIR reply. Please let me know if you have done this function.

**Comment by sga001** [comment]
You can use Kinan's version above.

I had implemented deserialization a long time ago, but it was for an older version of SealPIR.
